src/nvim/misc1.c

```diff
@@ -1059,3 +1059,33 @@ void add_time(char_u *buf, size_t buflen, time_t tt)
                  seconds);
   }
 }
+
+/// Fires a ModeChanged autocmd.
+void trigger_modechanged(void)
+{
+  if (!has_event(EVENT_MODECHANGED)) {
+    return;
+  }
+
+  char *mode = get_mode();
+  if (STRCMP(mode, last_mode) == 0) {
+    xfree(mode);
+    return;
+  }
+
+  dict_T *v_event = get_vim_var_dict(VV_EVENT);
+  tv_dict_add_str(v_event, S_LEN("new_mode"), mode);
+  tv_dict_add_str(v_event, S_LEN("old_mode"), last_mode);
+
+  char_u *pat_pre = concat_str((char_u *)last_mode, (char_u *)":");
+  char_u *pat = concat_str(pat_pre, (char_u *)mode);
+  xfree(pat_pre);
+
+  apply_autocmds(EVENT_MODECHANGED, pat, NULL, false, curbuf);
+  STRCPY(last_mode, mode);
+
+  xfree(pat);
+  tv_dict_free_contents(v_event);
+  hash_init(&v_event->dv_hashtab);
+  xfree(mode);
```

Comment:
Hmm, I think this is what janlazo wanted you to do:
```suggestion
  xfree(last_mode);
  last_mode = mode;

  xfree(pat);
  tv_dict_free_contents(v_event);
  hash_init(&v_event->dv_hashtab);
```

Reply:
Thanks, I only did the equivalent change in `autocmd.c` and totally forgot about the main part. This is done now and squashed into the port of 8.2.3430 (the first commit in this series)